README.md
#scribble

Turn a canvas element into a scribble pad that supports bouth mouse and touch.

##Installation

###Bower

`bower install scribble`

###Manual Download

- [Development]()
- [Production]()

##Usage

In order to initialize the module, you'll need to target either a `canvas` or a `div`. The module will wrap the `canvas` with a `div` or create a `canvas` inside the supplied `div`.

Finally, you need to just call `scribble` on top of the jQuery element:

	$('#myAwesomeCanvas').scribble();
	
And that's it! Well, I lied a little. You have some methods and options available!

### Available options

| Option            | Explanation                             | Default     |
| ----------------- |:---------------------------------------:|:-----------:|
| `color`           | Color in which you'll draw              | `#000000`   |
| `size`            | Size of the stroke                      | `2`         |
| `readMode`        | Sets whether the canvas is in read mode | `false`     |
| `tool`            | Selected tool at beginning              | `pencil`    |
| `stopDrawingTime` | Time to debounce stop drawing           | `500`       |
| `cssClasses`      | Object that holds css classes to style  | *See below* |

`cssClasses` have this configurable properties:

* `canvas-holder` : Class for the `div` that wraps the canvas. (*Default*: `scribble-canvas-holder`)
* `main-canvas` : Class for the main `canvas`. (*Default*: `scribble-main-canvas`)
* `shadow-canvas` : Class for the auxiliar canvas used. (*Default*: `scribble-main-canvas`)

**Note**: Aside from `cssClases`, all the options can be change with methods after initialization. 

### Available methods

Several methods are exposed through Fidel for you to use that allows you to do fancy things.

#### `changeColor(color)`

With this method you can… change the stroke color. What did you expect? There is no validation performed on plugin side so you should be sure you're passing a good value.

#### `changeSize(size)`

Changes the size of the stroke. Size has to be an integer.

#### `changeReadMode(mode)`

Changes the read mode to whatever value you pass. If you pass `true`, you'll enable read mode and `false` will disable it.

#### `undo`

This method will undo the latest drawing.

#### `redo`

This method will redo the latest undone action.

#### `clear`

This method will clear the drawing canvas.

#### `changeTool(tool)`

This method allows you to change "drawing" tool. Currently there are only two possible values. Plugin will yell if you pass something which is not implemented.

* `pencil` : This is the default value and is what you expect.
* `eraser` : This is what you expect too… It will erase as you draw.

### Fired events

#### `drawing.changed`

Whenever the user finished drawing, this event will fire. It's *debounced* so it won't fire until the time defined on the option `stopDrawingTime` passess without any more alike events.


#### Export methods

There are 2 exports methods with their counterpart that will import as well:

* `toJSON` : It returns an object literall with all the points.
* `loadJSON` : It loads that object and draws everything back.
* `toDataURL` : Converts the canvas to base64 string
* `loadDataUrl` : Loads the canvas from a base64 string

##Development

###Requirements

- node and npm
- bower `npm install -g bower`
- grunt `npm install -g grunt-cli`

###Setup

- `npm install`
- `bower install`

###Run

`grunt dev`

or for just running tests on file changes:

`grunt ci`

###Tests

`grunt mocha`


HISTORY.md

0.3.0 / 2014-05-30 
==================

  * Adding retina canvas support
  * New fullSteps feature that reduces the size of the points saved. Great if you are not changing tools, size and color on your scribble.
  * The loadDataURL method now uses asynchronous loading of the image which solves an issue of images not loading on some browsers  

0.2.0 / 2014-04-02 
==================

  * Removing console from test
  * Updating mocks to reflect new behaviour
  * Taking offset into account on touch devices and avoiding saving too many points
  * Updating jQuery to point to the new location

0.1.0 / 2013-09-21
==================
  * Fixes #5

0.0.1 / 2013-09-09 
==================

  * Fixes #4
  * Improve test coverage & fixes
  * Fix #3
  * Fixes #2
  * Adding Readme and some tests
  * First commit
  * project setup


test/scribble.test.js
function cleanMess(){
  $('#fixture').empty().html('<canvas id="testCanvas"></canvas><div id="testDiv"></div>');
}

function startDrawing(fidel,x,y){
  var down = $.Event('mousedown');

  down.layerX = x;
  down.layerY = y;

  movePoint(fidel,x,y);
  fidel.shadowCanvas.trigger(down);
}

function stopDrawing(fidel){
  var up = $.Event('mouseup');
  fidel.shadowCanvas.trigger(up);
}

function drawLine1(fidel){
  startDrawing(fidel,0,0);

  for (var i = 1; i < 5; i++){
    movePoint(fidel,i,i);
  }
  stopDrawing(fidel);

  return fidel.toJSON();
}

function movePoint(fidel,x,y){
  var move = $.Event('mousemove');

  move.layerX = x;
  move.layerY = y;

  fidel.shadowCanvas.trigger(move);
}

suite('scribble', function() {
  var canvas, div, canvasFidel, scribble;
  var line1Mock = [{"points":[{"x":0,"y":0,"size":2,"color":"#000000","tool":"pencil"},{"x":1,"y":1,"size":2,"color":"#000000","tool":"pencil"},{"x":2,"y":2,"size":2,"color":"#000000","tool":"pencil"},{"x":3,"y":3,"size":2,"color":"#000000","tool":"pencil"},{"x":4,"y":4,"size":2,"color":"#000000","tool":"pencil"}]}];

  setup(function(){
    cleanMess();
    canvas = $('#testCanvas');
    div = $('#testDiv');
    canvas.scribble();
    canvasFidel = canvas.data('scribble');
    div.scribble();
  });
  suite('init',function(){
    test('scribble should exists in jQuery\'s namespace', function(){
      assert.equal(typeof $().scribble, 'function');
    });
    suite('structure creating',function(){
      suite('canvas attributes',function(){
        test('main canvas should have id attribute',function(){
          assert.notEqual(div.find('canvas').first().attr('id'),'');
        });
        test('last canvas should have id attribute',function(){
          assert.notEqual(div.find('canvas').last().attr('id'),'');
        });
        test('main canvas should have width attribute',function(){
          assert.notEqual(div.find('canvas').first().attr('width'),'');
        });
        test('last canvas should have width attribute',function(){
          assert.notEqual(div.find('canvas').last().attr('width'),'');
        });
        test('main canvas should have height attribute',function(){
          assert.notEqual(div.find('canvas').first().attr('height'),'');
        });
        test('last canvas should have height attribute',function(){
          assert.notEqual(div.find('canvas').last().attr('height'),'');
        });
      });
      suite('canvas target',function(){
        test('scribble should create a wrapper div', function(){
          assert.ok(canvas.parent().hasClass(canvasFidel.cssClasses['canvas-holder']));
        });
        test('scribble should create a shadow canvas next to the target one', function(){
          assert.ok(canvas.next().hasClass(canvasFidel.cssClasses['shadow-canvas']));
        });
        test('scribble should assign the correct class to the canvas',function(){
          assert.ok(canvas.hasClass(canvasFidel.cssClasses['main-canvas']));
        });
      });
      suite('div target',function(){
        test('scribble should two canvas inside', function(){
          assert.equal(div.find('canvas').length,2);
        });
        test('scribble should add class to div', function(){
          assert.ok(div.hasClass(canvasFidel.cssClasses['canvas-holder']));
        });
        test('scribble should create a shadow canvas next to the target one', function(){
          assert.equal(div.find('canvas.' +canvasFidel.cssClasses['shadow-canvas']).length,1);
        });
        test('scribble should assign the correct class to the canvas',function(){
          assert.equal(div.find('canvas.' +canvasFidel.cssClasses['main-canvas']).length,1);
        });
      });
    });
  });
  suite('feature',function(){
    setup(function(){
      cleanMess();
      canvas = $('#testCanvas');
      div = $('#testDiv');
      canvas.scribble();
      canvasFidel = canvas.data('scribble');
      div.scribble();
    });
    suite('drawing',function(){
      setup(function(){
        cleanMess();
        canvas = $('#testCanvas');
        scribble = canvas.scribble();
        canvasFidel = canvas.data('scribble');
        drawLine1(canvasFidel);
      });
      test('scribble should draw a line when user tries to',function(){
        assert.deepEqual(canvasFidel.toJSON(),line1Mock);
      });
      test('scribble should undo the drawing when calling undo',function(){
        canvasFidel.undo();
        assert.deepEqual(canvasFidel.toJSON(),[]);
      });
      test('scribble should redo the drawing when calling redo',function(){
        canvasFidel.undo();
        canvasFidel.redo();
        assert.deepEqual(canvasFidel.toJSON(),line1Mock);
      });
      test('scribble should empty the canvas on clear',function(){
        canvasFidel.clear();
        assert.deepEqual(canvasFidel.toJSON(),[]);
      });
      test('scribble should fire an event when line was drawn', function(done){
        scribble.on('drawing.changed',function(){
          done();
        });
      });
      test('scribble should save simpler points if fullSteps is false', function () {
        cleanMess();
        canvas = $('#testCanvas');
        scribble = canvas.scribble({ fullSteps : false });
        canvasFidel = canvas.data('scribble');
        drawLine1(canvasFidel);
        var firstPoint = canvasFidel.toJSON()[0].points[0];

        assert.ok(typeof firstPoint.color === "undefined");
        assert.ok(typeof firstPoint.size === "undefined");
        assert.ok(typeof firstPoint.tool === "undefined");
        assert.ok(typeof firstPoint.x !== "undefined");
        assert.ok(typeof firstPoint.y !== "undefined");
      });
    });
    suite('exporting',function(){
      setup(function(){
        cleanMess();
        canvas = $('#testCanvas');
        canvas.scribble();
        canvasFidel = canvas.data('scribble');
        drawLine1(canvasFidel);
      });
      test('scribble should export drawing to JSON',function(){
        assert.deepEqual(canvasFidel.toJSON(),line1Mock);
      });
      test('scribble should export drawing to base64',function(){
        assert.ok(canvasFidel.toDataURL().indexOf('data:image/png;base64') !== -1);
      });
    });
    suite('importing',function(){
      setup(function(){
        cleanMess();
        canvas = $('#testCanvas');
        canvas.scribble();
        canvasFidel = canvas.data('scribble');
      });
      test('scribble should import drawing from JSON',function(){
        canvasFidel.loadJSON(line1Mock);
        assert.deepEqual(canvasFidel.toJSON(),line1Mock);
      });
    });
  });
});


Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    info: grunt.file.readJSON('bower.json'),
    meta: {
      banner: '/*!\n'+
              ' * <%= info.name %> - <%= info.description %>\n'+
              ' * v<%= info.version %>\n'+
              ' * <%= info.homepage %>\n'+
              ' * copyright <%= info.copyright %> <%= grunt.template.today("yyyy") %>\n'+
              ' * <%= info.license %> License\n'+
              '*/\n'
    },
    jshint: {
      main: [
        'Gruntfile.js',
        'bower.json',
        'lib/**/*.js',
        'test/*.js'
      ]
    },
    bower: {
      main: {
        dest: 'dist/_bower.js',
        exclude: [
          'jquery',
          'assert'
        ]
      }
    },
    concat: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        src: [
          'lib/scribble.js'
        ],
        dest: 'dist/fidel.scribble.js'
      },
      full: {
        src: [
          'dist/_bower.js',
          'bower_components/hidpi-canvas/dist/hidpi-canvas.js',
          'lib/scribble.js'
        ],
        dest: 'dist/scribble.js'
      }
    },
    uglify: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        src: 'dist/fidel.scribble.js',
        dest: 'dist/fidel.scribble.min.js'
      },
      full: {
        src: 'dist/scribble.js',
        dest: 'dist/scribble.min.js'
      }
    },
    clean: {
      bower: [
        'dist/_bower.js'
      ],
      dist: [
        'dist'
      ]
    },
    watch: {
      scripts: {
        files: '<%= jshint.main %>',
        tasks: 'scripts',
        options: {
          livereload: true
        }
      },
      example: {
        files: [
          'example/*'
        ],
        options: {
          livereload: true
        }
      },
      grunt: {
        files: [
          'Gruntfile.js',
          'test/index.html'
        ],
        tasks: 'default'
      }
    },
    mocha: {
      options: {
        run: true,
        growl: true,
        reporter: 'Spec'
      },
      all: {
        src: 'test/index.html'
      }
    },
    plato: {
      main: {
        files: {
          'reports': ['lib/*.js']
        }
      }
    },
    connect: {
      server:{
        options: {
          hostname: '*'
        }
      },
      plato: {
        port: 8000,
        base: 'reports',
        options: {
          keepalive: true
        }
      }
    },
    bytesize: {
      scripts: {
        src: [
          'dist/*'
        ]
      }
    }
  });
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-connect');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-concat-bower');
  grunt.loadNpmTasks('grunt-shell');
  grunt.loadNpmTasks('grunt-bytesize');
  grunt.loadNpmTasks('grunt-mocha');
  grunt.loadNpmTasks('grunt-plato');
  grunt.registerTask('scripts', ['jshint', 'bower', 'concat', 'uglify', 'clean:bower', 'mocha', 'bytesize']);
  grunt.registerTask('default', ['scripts']);
  grunt.registerTask('dev', ['connect:server', 'watch']);
  grunt.registerTask('reports', ['plato', 'connect:plato']);
};


lib/scribble.js
(function($){
  function capitalize (string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
  }

  $.declare('scribble',{
    defaults : {
      color : '#000000',
      size : 2,
      readMode : false,
      stopDrawingTime : 500,
      fullSteps : true,
      tool : 'pencil',
      cssClasses : {
        'canvas-holder' : 'scribble-canvas-holder',
        'main-canvas' : 'scribble-main-canvas',
        'shadow-canvas' : 'scribble-shadow-canvas'
      }
    },
    tools : [
      'pencil','eraser'
    ],
    _isCanvas : function () {
      var htmlNode = this.el[0],
          result = true;

      if (htmlNode.nodeType !== 1 || htmlNode.nodeName.toLowerCase() !== 'canvas'){
        result = false;
      }

      return result;
    },
    _createCanvas : function(){
      var canvas = $('<canvas>');
      canvas.appendTo(this.el);
      this.canvasHolder = this.el;
      this.el = canvas;
    },
    _getId : function() {
      var i = 0,id = null, auxID;

      while (id === null) {
        auxID = 'fidel_scribble_' + i;

        if ($('#' + auxID).length === 0){
          id = auxID;
        }
        else {
          i++;
        }
      }

      return id;
    },
    _createWrapper : function(){
      var div = $('<div>');
      div.insertBefore(this.el);
      this.el.appendTo(div);
      this.canvasHolder = div;
    },
    _applyClasses : function(){
      this.canvasHolder.addClass(this.cssClasses['canvas-holder']);
      this.el.addClass(this.cssClasses['main-canvas']);
      this.shadowCanvas.addClass(this.cssClasses['shadow-canvas']);
      var computedStyle = getComputedStyle(this.canvasHolder[0]);

      this.el[0].width = parseInt(computedStyle.width,10);
      this.el[0].height = parseInt(computedStyle.height,10);
      this.shadowCanvas[0].width = parseInt(computedStyle.width,10);
      this.shadowCanvas[0].height = parseInt(computedStyle.height,10);
    },
    _getContexts : function(){
      this.shadowContext = this.shadowCanvas[0].getContext('2d');
      this.context = this.el[0].getContext('2d');
    },
    _createShadowCanvas : function(){
      this.shadowCanvas = $('<canvas>');
      this.shadowCanvas[0].id = this._getId();
      this.shadowCanvas.insertAfter(this.el);
    },
    _bindEvents: function () {
      this.shadowCanvas.on('mousemove touchmove', this.proxy(this._saveMouse,this));
      this.shadowCanvas.on('mousedown touchstart', this.proxy(this._startDrawing,this));
      this.shadowCanvas.on('mouseup touchend', this.proxy(this._stopDrawing,this));

      $('body').on('mouseup touchend', this.proxy(this._stopDrawing,this));
    },
    _unbindEvents : function(){
      this.shadowCanvas.off('mousemove touchmove', this.proxy(this._saveMouse,this));
      this.shadowCanvas.off('mousedown touchstart', this.proxy(this._startDrawing,this));
      this.shadowCanvas.off('mouseup touchend', this.proxy(this._stopDrawing,this));

      $('body').off('mouseup', this.proxy(this._stopDrawing,this));
    },
    _startDrawing : function(e){
      e.stopPropagation();
      e.preventDefault();

      if(e.handled !== true) {
        this.shadowCanvas.on('mousemove touchmove', this.proxy(this._draw, this));
        this.drawing = true;
        this._saveMouse(e);
        this._draw();
        e.handled = true;
      }
      else {
        return false;
      }
    },
    _savePoint : function(){
      var point = {
        x : this.mousePosition.x,
        y : this.mousePosition.y
      };

      if (this.fullSteps){
        point.size = this.size;
        point.color = this.color;
        point.tool = this.tool;
      }

      this.points.push(point);
    },
    _saveMouse : function(e){
      if (e.type === 'touchmove'){
        e.preventDefault();
      }

      var position = this._getXY(e);

      this.mousePosition.x = position.x;
      this.mousePosition.y = position.y;

      if (this.drawing){
        this._savePoint();
      }
    },
    _saveStep : function(){
      var aux = { points: this.points.splice(0) };
      this.doneSteps.push(aux);
    },
    _stopDrawing : function(e){
      if (this.drawing){
        e.stopImmediatePropagation();
        this.shadowCanvas.off('mousemove touchmove', this.proxy(this._draw,this));
        this._copyShadowToReal();
        this.drawing = false;
        this._emitDrawingChanged();
      }
    },
    _emitDrawingChanged : function(){
      if (this.stopTimer){
        clearTimeout(this.stopTimer);
      }
      var self = this;
      this.stopTimer = setTimeout(function(){
        self.emit('drawing.changed');
      }, this.stopDrawingTime);
    },
    _copyShadowToReal : function(){
      this.context.drawImage(this.shadowCanvas[0],0,0);
      this._clearCanvas(this.shadowCanvas,this.shadowContext);
      if (this.drawing){
        this._saveStep();
      }
      this.points = [];
    },
    _getXY : function(e){
      var x, y, touchEvent = { pageX : 0, pageY : 0};

      if (typeof e.changedTouches !== 'undefined'){
        touchEvent = e.changedTouches[0];
      }
      else if ( typeof e.originalEvent !== 'undefined' && typeof e.originalEvent.changedTouches !== 'undefined'){
        touchEvent = e.originalEvent.changedTouches[0];
      }

      x = e.offsetX || e.layerX || touchEvent.pageX;
      y = e.offsetY || e.layerY || touchEvent.pageY;

      if (e.type.indexOf('touch') !== -1){
        var offset = this.shadowCanvas.offset();

        x -= offset.left;
        y -= offset.top;
      }

      return {
        x : x,
        y : y
      };
    },
    _draw : function(){
      var length = this.points.length,
          aux = this.points[0],
          erasing = aux.tool === 'eraser',
          context =  erasing ? this.context : this.shadowContext,
          previousTool = '',
          revertTool = false;

      if (this.tool !== aux.tool && typeof aux.tool != "undefined"){
        revertTool = true;
        previousTool = this.tool;
        this.changeTool(aux.tool);
      }

      context.lineWidth = aux.size || this.size;
      context.strokeStyle = aux.color || this.color;
      context.fillStyle = aux.color || this.color;

      if (length !== 0){
        if (length < 3){
          context.beginPath();
          context.arc(aux.x,aux.y, context.lineWidth / 2, 0, Math.PI * 2, true);
          context.fill();
          context.closePath();
        }
        else {
          if (!erasing){
            this._clearCanvas(this.shadowCanvas,this.shadowContext);
          }

          context.beginPath();
          context.moveTo(this.points[0].x,this.points[0].y);

          for (var i = 1; i < length -2; i++){
            var x = (this.points[i].x + this.points[i+1].x) / 2,
                y = (this.points[i].y + this.points[i+1].y) / 2;

            context.quadraticCurveTo(this.points[i].x,this.points[i].y,x,y);
          }

          context.quadraticCurveTo(this.points[i].x,this.points[i].y,this.points[i+1].x,this.points[i+1].y);
          context.stroke();
        }
      }

      // Only used in undo / redo
      return {
        'toolUsed' : aux.tool,
        'previousTool' : previousTool,
        'revertTool' : revertTool
      };
    },
    _drawFromSteps : function(actions){
      if (actions){
        this.clear(true);

        for (var i = 0, len = actions.length; i < len; i++){
          this.points = actions[i].points;
          var result = this._draw();

          if (result.toolUsed !== 'eraser'){
            this._copyShadowToReal();
          }
          if (result.revertTool){
            this.changeTool(result.previousTool);
          }
        }
      }
    },
    init : function(){
      if (!this._isCanvas()){
        this._createCanvas();
      }
      else {
        this._createWrapper();
      }
      this._createShadowCanvas();

      this._applyClasses();
      this._getContexts();

      this.shadowContext.lineJoin = 'round';
      this.body = $('body');

      this.mousePosition = {
        x : 0,
        y : 0
      };
      this.points = [];
      this.drawing = false;
      this.unDoneSteps = [];
      this.doneSteps = [];

      this.clear(true);

      if (!this.readMode){
        this._bindEvents();
      }
      this.oldColor = this.color;
      this.changeTool(this.tool);
    },
    changeColor: function(color){
      this.color = color;
      this.shadowContext.strokeStyle = this.color;
    },
    changeSize : function(size){
      this.size = size;
      this.shadowContext.lineWidth = this.size;
    },
    changeReadMode : function(mode){
      if (mode){
        this._bindEvents();
      }
      else {
        this._unbindEvents();
      }

      this.readMode = mode;
    },
    toJSON : function(){
      return this.doneSteps;
    },
    loadJSON : function(obj){
      this._drawFromSteps(obj);
      this.doneSteps = obj;
    },
    toDataURL : function(){
      return this.el[0].toDataURL();
    },
    loadDataURL : function(dataurl){
      var image = new Image(),
          context = this.context;

      image.onload = function() {
        context.drawImage(this, 0, 0);
      };

      image.src = dataurl.toString();
    },
    undo : function(){
      if (this.doneSteps.length){
        var lastStep = this.doneSteps.pop();
        this.unDoneSteps.push(lastStep);
        this._drawFromSteps(this.doneSteps);
      }
    },
    redo : function(){
      if (this.unDoneSteps.length){
        var lastUndone = this.unDoneSteps.pop();
        this.doneSteps.push(lastUndone);
        this._drawFromSteps(this.doneSteps);
      }
    },
    _setPencil : function(){
      this.context.globalCompositeOperation = 'source-over';
      this.color = this.oldColor;
    },
    _setEraser : function(){
      this.context.globalCompositeOperation = 'destination-out';
      this.oldColor = this.color;
      this.color = 'rgba(0,0,0,1)';
    },
    changeTool : function(tool){
      if (this.tools.indexOf(tool) !== -1){
        var method = '_set' + capitalize(tool);
        this[method].call(this);
        this.tool = tool;
      }
      else {
        console.error(tool + ' is not implemented');
      }
    },
    _clearCanvas : function(canvas, context){
      context.clearRect(0,0,canvas[0].width,canvas[0].height);
    },
    clear : function(intern){
      intern = intern || false;

      this._clearCanvas(this.el,this.context);

      if (!intern){
        this.emit('clear');
        this.doneSteps = [];
      }
    }
  });
})(jQuery);